src/fluree/raft/log.clj

```diff
@@ -67,38 +76,169 @@
+    (log/warn (str msg error-at known-info))))
+
+(defn move-log
+  "Moves a log file from the source-path to destination-path"
+  [source-path destination-path]
+  (let [source-uri      (URI/create (str "file://" source-path))
+        destination-uri (URI/create (str "file://" destination-path))
+        source          (Paths/get source-uri)
+        destination     (Paths/get destination-uri)]
+    (Files/move source destination
+                (into-array CopyOption [StandardCopyOption/ATOMIC_MOVE
+                                        StandardCopyOption/REPLACE_EXISTING]))))
+
+(defn copy-corrupt-file
+  "Creates a copy (for forensics) of a corrupt raft file using the '.corrupt. extension"
+  [^File file]
+  (let [source-path      (.getPath file)
+        destination-path (str source-path ".corrupt")
+        destination      (io/file destination-path)]
+    (log/info (str "Copying corrupted log file for forensics from: " source-path " to: " destination-path))
+    (io/copy file destination)
+    (log/debug (str "Successfully created file: " destination-path))
+    :done))
+
+(defn repair-file
+  "When we have an end of file exception due to a corrupt final log entry,
+  we need to fix the file so that the log can continue to be used."
+  [^File file log]
+  (try
+    (log/debug (str "About to repair log file: " (.getName file) " with " (count log) " valid entries."))
+    (let [source-path (.getPath file)
+          repair-path (str source-path ".repaired")
+          repair-file (io/file repair-path)]
+      (copy-corrupt-file file)
+      (log/debug (str "Repairing log file: Writing new correct log to: " repair-path " as a temporary file."))
+      (write-log repair-file log)
+      (log/debug (str "Repairing log file: Done writing temporary file: " repair-path
+                      ". Moving temporary file to replace original corrupt log: " source-path))
+      (move-log repair-file source-path)
+      true)
+    (catch Exception e
+      (throw (ex-info (str "Error creating repaired raft log file: " (.getName file)
+                           " with exception: " (ex-message e))
+                      {:file-name (.getName file)}
+                      e)))))
+
+(defn throw-corrupt-file-exception
+  "When reading a raft log file, report out a consistent format for exceptions if they occur."
+  [message exception ^RandomAccessFile raf file-name next-bytes index term entry-type entry-data]
+  (let [file-length  (.length raf)
+        file-pointer (.getFilePointer raf)
+        EOF?         (= file-length file-pointer)]
+    (throw (ex-info message
+                    {:status                 500
+                     :error                  :raft/corrupt-log
+                     :file-name              file-name
+                     :file-length            file-length
+                     :file-pointer           file-pointer
+                     :end-of-file-exception? EOF?
```

Comment:
could we get this info by reading whether the exception below is an `EOFException` with `ex-cause`?

Reply:
Yes, I think so. Let me see what that looks like.